File: AIIO 2008/DivaIIIANoteGoneSour.cpp
// https://orac2.info/problem/144/
#include <bits/stdc++.h>
using namespace std;
 
int main() {
    ifstream fin("divain.txt");
    ofstream fout("divaout.txt");
 
    int N, T, K;
    fin >> N >> T;
    fin >> K;
 
    map<int, vector<pair<int, int>>> events_by_day;  // day -> list of (musician, effect)
 
    for (int i = 0; i < K; ++i) {
        int d, m, r;
        fin >> d >> m >> r;
        events_by_day[d].emplace_back(m, r);
    }
 
    vector<int> current_effect(N + 1, 0);  // current slope for each musician
    vector<int> last_effect(N + 1, 0);     // effect before the latest change
 
    long long current_quality = 0;
    long long slope = 0;
    int last_day = 0;
 
    // Store all unique days to process (events + T + 1 to close the interval)
    set<int> important_days;
    for (const auto& [day, _] : events_by_day) important_days.insert(day);
    important_days.insert(T + 1);  // to finalize till end
 
    int best_day = 0;
    long long best_quality = 0;
 
    for (int day : important_days) {
        int delta_days = day - last_day;
        current_quality += slope * delta_days;
 
        // If equal quality, prefer the LATEST such day
        if (current_quality >= best_quality) {
            best_quality = current_quality;
            best_day = day - 1;
        }
 
        // Apply all effect changes at current day
        for (auto& [musician, new_r] : events_by_day[day]) {
            slope -= current_effect[musician];
            current_effect[musician] = new_r;
            slope += current_effect[musician];
        }
 
        last_day = day;
    }
 
    fout << best_day << " " << best_quality << '\n';
    return 0;
}

// approach was APPARENTLY too memory inneficient in python

File: AIIO 2008/DivaIIIANoteGoneSour.py
# https://orac2.info/problem/144/
# this is too memory inneficient on 3 testcases, check the C++ file in the same directory as this file
import sys
from collections import defaultdict

sys.stdin = open('divain.txt', 'r')
sys.stdout = open('divaout.txt', 'w')

idx = 0
N, T = map(int, input().split())
idx += 1
K = int(input())
idx += 1

events_by_day = defaultdict(list)  # day -> list of (musician, effect)

for _ in range(K):
    d, m, r = map(int, input().split())
    events_by_day[d].append((m, r))
    idx += 1

current_effect = [0] * (N + 1)  # current slope for each musician
slope = 0
current_quality = 0
last_day = 0

# Store all unique days to process (events + T + 1 to close the interval)
important_days = set(events_by_day.keys())
important_days.add(T + 1)  # to finalize till end
important_days = sorted(important_days)

best_day = 0
best_quality = 0

for day in important_days:
    delta_days = day - last_day
    current_quality += slope * delta_days

    # If equal quality, prefer the LATEST such day
    if current_quality >= best_quality:
        best_quality = current_quality
        best_day = day - 1

    # Apply all effect changes at current day
    for musician, new_r in events_by_day.get(day, []):
        slope -= current_effect[musician]
        current_effect[musician] = new_r
        slope += current_effect[musician]

    last_day = day

print(f"{best_day} {best_quality}\n")
